src/dbnode/persist/fs/persist_manager.go

```diff
@@ -115,22 +117,100 @@ type dataPersistManager struct {
 	snapshotID uuid.UUID
 }
 
-type indexPersistManager struct {
-	writer        IndexFileSetWriter
-	segmentWriter m3ninxpersist.MutableSegmentFileSetWriter
+// Support writing to multiple index blocks/filesets during index persist.
+// This allows us to prepare an index fileset writer per block start.
+type singleUseIndexWriter struct {
+	// back-ref to the index persist manager so we can share resources there
+	manager *indexPersistManager
+	writer  IndexFileSetWriter
 
 	// identifiers required to know which file to open
 	// after persistence is over
 	fileSetIdentifier FileSetFileIdentifier
 	fileSetType       persist.FileSetType
 
-	// track state of writers
-	writeErr    error
-	initialized bool
+	// track state of writer
+	writeErr error
+}
+
+func (s *singleUseIndexWriter) persistIndex(builder segment.Builder) error {
+	// Lock the index persist manager as we're sharing the segment builder as a resource.
+	s.manager.Lock()
+	defer s.manager.Unlock()
+
+	markError := func(err error) {
+		s.writeErr = err
+	}
+	if err := s.writeErr; err != nil {
+		return fmt.Errorf("encountered error: %v, skipping further attempts to persist data", err)
+	}
+
+	if err := s.manager.segmentWriter.Reset(builder); err != nil {
+		markError(err)
+		return err
+	}
+
+	if err := s.writer.WriteSegmentFileSet(s.manager.segmentWriter); err != nil {
+		markError(err)
+		return err
+	}
+
+	return nil
+}
+
+func (s *singleUseIndexWriter) closeIndex() ([]segment.Segment, error) {
```

Comment:
Maybe also hold the `s.manager.Lock` for same level of lock granularity as the persistIndex method? Even if not required today, someone might add something to the close method that could be used concurrently by mistake later.